### ISSUE TITLE
synthseg crash

### ISSUE THREAD

**Issue by marcelzwiers** [PennLINC/qsiprep#595]
I'm running qsiprep 0.18.0 on a standard (bids-validated) dataset like this:
```console
tree bids/sub-001
bids/sub-001
|-- anat
|   |-- sub-001_acq-MPRAGE_rec-biascorr_run-1_T1w.json
|   |-- sub-001_acq-MPRAGE_rec-biascorr_run-1_T1w.nii.gz
|   |-- sub-001_acq-TSE_rec-biascorr_run-1_T2w.json
|   |-- sub-001_acq-TSE_rec-biascorr_run-1_T2w.nii.gz
|   |-- sub-001_run-1_UNIT1.json
|   |-- sub-001_run-1_UNIT1.nii.gz
|   |-- sub-001_run-1_echo-1_part-mag_MEGRE.json
|   |-- sub-001_run-1_echo-1_part-mag_MEGRE.nii.gz
|   |-- sub-001_run-1_echo-1_part-phase_MEGRE.json
|   |-- sub-001_run-1_echo-1_part-phase_MEGRE.nii.gz
|   |-- [..]
|   |-- sub-001_run-1_echo-9_part-mag_MEGRE.json
|   |-- sub-001_run-1_echo-9_part-mag_MEGRE.nii.gz
|   |-- sub-001_run-1_echo-9_part-phase_MEGRE.json
|   |-- sub-001_run-1_echo-9_part-phase_MEGRE.nii.gz
|   |-- sub-001_run-1_inv-1_MP2RAGE.json
|   |-- sub-001_run-1_inv-1_MP2RAGE.nii.gz
|   |-- sub-001_run-1_inv-2_MP2RAGE.json
|   `-- sub-001_run-1_inv-2_MP2RAGE.nii.gz
|-- dwi
|   |-- sub-001_acq-HARDI_dir-AP_run-1_dwi.bval
|   |-- sub-001_acq-HARDI_dir-AP_run-1_dwi.bvec
|   |-- sub-001_acq-HARDI_dir-AP_run-1_dwi.json
|   |-- sub-001_acq-HARDI_dir-AP_run-1_dwi.nii.gz
|   |-- sub-001_acq-HARDI_dir-AP_run-1_sbref.json
|   `-- sub-001_acq-HARDI_dir-AP_run-1_sbref.nii.gz
`-- sub-001_scans.tsv

qsub -l nodes=1:ppn=2,walltime=72:00:00,mem=20000mb,file=50gb -N qsiprep_sub-001  <<EOF
cd /project/3015046.06
apptainer run --cleanenv  --bind \$TMPDIR:/tmp /opt/qsiprep/0.18.0/qsiprep-0.18.0.sif bids bids/derivatives participant --output-resolution 1.8 -w \$TMPDIR/sub-001 --participant-label 001 --skip-bids-validation --fs-license-file /opt_host/fmriprep/license.txt --mem_mb 20000 --omp-nthreads 2 --nthreads 2 
EOF
```
It runs for several hours but crashes on mri_synthseg:
```console
QSIPrep failed: Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node synthseg.

Cmdline:
        mri_synthseg --i /scratch/marzwi/49347907.dccn-l029.dccn.nl/sub-260/qsiprep_wf/single_subject_260_wf/anat_preproc_wf/pad_anat_reference_wf/resample_skulled_to_reference/sub-260_acq-MPRAGE_rec-biascorr_run-1_T1w_lps_trans.nii.gz --threads 2 --post sub-260_acq-MPRAGE_rec-biascorr_run-1_T1w_lps_trans_post.nii.gz --qc sub-260_acq-MPRAGE_rec-biascorr_run-1_T1w_lps_trans_qc.csv --o sub-260_acq-MPRAGE_rec-biascorr_run-1_T1w_lps_trans_aseg.nii.gz
Stdout:

Stderr:
        Aborted (core dumped)
Traceback:
        Traceback (most recent call last):
          File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 453, in aggregate_outputs
            setattr(outputs, key, val)
          File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/traits_extension.py", line 330, in validate
            value = super(File, self).validate(objekt, name, value, return_pathlike=True)
          File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/traits_extension.py", line 135, in validate
            self.error(objekt, name, str(value))
          File "/usr/local/miniconda/lib/python3.8/site-packages/traits/base_trait_handler.py", line 74, in error
            raise TraitError(
        traits.trait_errors.TraitError: The 'out_seg' trait of a _SynthSegOutputSpec instance must be a pathlike object or string representing an existing file, but a value of '/scratch/marzwi/49347907.dccn-l029.dccn.nl/sub-260/qsiprep_wf/single_subject_260_wf/anat_preproc_wf/synthseg_anat_wf/synthseg/sub-260_acq-MPRAGE_rec-biascorr_run-1_T1w_lps_trans_aseg.nii.gz' <class 'str'> was specified.

        During handling of the above exception, another exception occurred:
[..]
```
See the attached files for more details. I'm stuck here, does anybody have any suggestions on what to do now?

[qsiprep_sub-001.e49347653.log](https://github.com/PennLINC/qsiprep/files/11892560/qsiprep_sub-001.e49347653.log)
[qsiprep_sub-001.o49347653.log](https://github.com/PennLINC/qsiprep/files/11892561/qsiprep_sub-001.o49347653.log)
[crash-20230627-123725-marzwi-synthseg-859f4773-2ab1-45b4-8376-7d61b7d426c3.txt](https://github.com/PennLINC/qsiprep/files/11892569/crash-20230627-123725-marzwi-synthseg-859f4773-2ab1-45b4-8376-7d61b7d426c3.txt)

**Comment by marcelzwiers** [comment]
[crash-20230627-123725-marzwi-synthseg-859f4773-2ab1-45b4-8376-7d61b7d426c3.txt](https://github.com/PennLINC/qsiprep/files/11892617/crash-20230627-123725-marzwi-synthseg-859f4773-2ab1-45b4-8376-7d61b7d426c3.txt)
[qsiprep_sub-001.e49347653.log](https://github.com/PennLINC/qsiprep/files/11892629/qsiprep_sub-001.e49347653.log)
[qsiprep_sub-001.o49347653.log](https://github.com/PennLINC/qsiprep/files/11892630/qsiprep_sub-001.o49347653.log)

**Comment by mattcieslak** [comment]
I would bet on there not being enough memory. SynthSeg takes a massive amount of memory to run, and if it's also being run in parallel with other things (here you have a very modest 2 threads). If possible I would increase memory to ~40GB just to be safe. Or you can fix nthreads to 1, which should work with the 20GB you're requesting but will take a long time to run.

**Comment by marcelzwiers** [comment]
I got the same error using 40GB (and 2 threads)...

**Comment by marcelzwiers** [comment]
Maybe it helps to mention that the voxelsize of my T1w image is 0.8mm^3 (which was not a problem for qsiprep 0.16)

**Comment by cookpa** [comment]
Synthseg should resample the image to 1mm isotropic. But I looked at the `mri_synthseg` script and it does not appear to crop the image by default, which the original synthseg implementation does. This would mean its memory requirements will scale by the size of the T1w image FOV.

**Comment by mattcieslak** [comment]
@marcelzwiers do you still have a copy of the working directory from the crashing runs? 

QSIPrep does its own resampling of the images using ants to avoid the internal image processing in the synths (which was causing even MORE memory use), maybe there is a bug in how this workflow handles <1mm inputs.

**Comment by cni-md** [comment]
I think I have the same problem with singularity. I have 512 GB of mem.

singularity run --cleanenv -B /license:/license -B ${TMPDIR}:/tmp -B ${NIFTI}:/input -B {OUTPUT}:/output  /pool/container/qsiprep_v0.18.1_js.simg  --participant-label jg45  --freesurfer-input /output/fmriprep/sourcedata/freesurfer --output-resolution 2 --hmc-model eddy --ignore fieldmaps --use-syn-sdc --force_syn  --fs-license /license/license --separate_all_dwis --nthreads 1  --omp-nthreads 1 -w /tmp  --notrack /input /output participant

nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node synthseg.

Cmdline:
	mri_synthseg --i /tmp/qsiprep_wf/single_subject_jg45_wf/anat_preproc_wf/pad_anat_reference_wf/resample_skulled_to_reference/sub-jg45_ses-20230627_acq-mprage10p2_T1w_lps_trans.nii.gz --threads 1 --post sub-jg45_ses-20230627_acq-mprage10p2_T1w_lps_trans_post.nii.gz --qc sub-jg45_ses-20230627_acq-mprage10p2_T1w_lps_trans_qc.csv --o sub-jg45_ses-20230627_acq-mprage10p2_T1w_lps_trans_aseg.nii.gz
Stdout:

Stderr:
	Illegal instruction
Traceback:
	Traceback (most recent call last):
	  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 453, in aggregate_outputs
	    setattr(outputs, key, val)
	  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/traits_extension.py", line 330, in validate
	    value = super(File, self).validate(objekt, name, value, return_pathlike=True)
	  File "/usr/local/miniconda/lib/python3.8/site-packages/nipype/interfaces/base/traits_extension.py", line 135, in validate
	    self.error(objekt, name, str(value))
	  File "/usr/local/miniconda/lib/python3.8/site-packages/traits/base_trait_handler.py", line 74, in error
	    raise TraitError(
	traits.trait_errors.TraitError: The 'out_post' trait of a _SynthSegOutputSpec instance must be a pathlike object or string representing an existing file, but a value of '/tmp/qsiprep_wf/single_subject_jg45_wf/anat_preproc_wf/synthseg_anat_wf/synthseg/sub-jg45_ses-20230627_acq-mprage10p2_T1w_lps_trans_post.nii.gz' <class 'str'> was specified.

**Comment by cookpa** [comment]
I've found that synthseg memory use peaks shortly before it exits - watching the process with htop it is stable for a while then shoots up.

**Comment by cni-md** [comment]
Are you able to call mri_synthseg within the container?
singularity shell   /pool/container/qsiprep_v0.18.1_js.simg

Singularity qsiprep_v0.18.1_js.simg:/opt/freesurfer/bin> mri_synthseg   
RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject
Illegal instruction

**Comment by cni-md** [comment]
Sorry, it was my modified container to avoid the bmat error. With the origiral container I run into 
ImportError: cannot import name 'bmat' from 'scipy.sparse.sputils' (/usr/local/miniconda/lib/python3.8/site-packages/scipy/sparse/sputils.py)

again.

**Comment by marcelzwiers** [comment]
Do you recommend `--containall` over `--cleanenv`? I always use the latter but I don't know about advantages of containall (except for disk mounts)...

**Comment by mattcieslak** [comment]
I've been using --containall recently, it will also keep the environment variables from being included. Lots of problems recently are from python libraries in home directories

**Comment by cni-md** [comment]
I tried singularity shell --containerall qsiprep_v0.18.1.simg (or --cleanall) 
/opt/freesurfer/bin/mri_synthstrip --help  is fine
/opt/freesurfer/bin/mri_synthseg --help
Illegal instruction

docker exec -it <my container id> bash
/opt/freesurfer/bin/mri_synthseg --help
Illegal instruction (core dumped)

The same on a different server with apptainer
/opt/freesurfer/bin/mri_synthseg --help
Illegal instruction

I am not able to call mri_synthseg --i my_file --o ... or the command mentioned in the log file.

Some of the other commands in /opt/freesurfer/bin are not working due to missing libraries.

**Comment by marcelzwiers** [comment]
What can happen is that singularity containers use python libraries from your `.local` folder in your home directory. You can avoid that by running `export SINGULARITY_PYTHONNOUSERSITE=1` before you run your container

**Comment by cookpa** [comment]
That's the nice thing about `--containall`, it prohibits the user home directory from being mounted, so it's not possible for any of the `~/.*` config directories to affect the container's behavior.

**Comment by cni-md** [comment]
@mattcieslak  sorry, typo. I used containall.
On buster I am getting always the bmat error with singularity 2.6.1-dist. With --clearall, --containall, wtih export SINGULARITY_PYTHONNOUSERSITE=1
With docker the traits error. 

Wtih bullseye (different server, different ~ dir) the traits error, 

The servers  are older ones with 80 cores  Intel(R) Xeon(R) CPU E7- 4870  @ 2.40GHz, 512 GB mem.

**Comment by mattcieslak** [comment]
@marcelzwiers I'm also seeing some issues with your BIDS curation in the error log. Have you tried getting the input data to pass the BIDS validator?

**Comment by marcelzwiers** [comment]
I'm on holiday now so just a quick answer. Yes, I did check with the
bids-validator, and that was all fine. It's a bids 1.8 dataset, what did
you see?

_____________________
Sent from my phone

Op vr 28 jul. 2023 16:50 schreef Matt Cieslak ***@***.***>:

**Comment by mattcieslak** [comment]
this is likely a problem with the bids validator in the qsiprep container. Sorry for the misdirection

**Comment by mattcieslak** [comment]
This should be fixed, or at least produce a more informative error in 0.20.0. Please open a new issue if there are still problems!
